velox/expression/benchmarks/VariadicBenchmark.cpp

```diff
@@ -0,0 +1,208 @@
+
+// Benchmark a function that concats its string inputs.
+
+namespace facebook::velox::functions {
+
+template <typename T>
+struct StringConcatSimple {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& out,
+      const arg_type<Varchar>& str1,
+      const arg_type<Varchar>& str2,
+      const arg_type<Varchar>& str3,
+      const arg_type<Varchar>& str4,
+      const arg_type<Varchar>& str5) {
+    stringImpl::concatStatic(out, str1, str2, str3, str4, str5);
+
+    return true;
+  }
+};
+
+template <typename T>
+struct StringConcatVariadic {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& out,
+      const arg_type<Variadic<Varchar>>& strings) {
+    for (const auto& string : strings) {
+      const auto& value = string.value();
+      if (value.size() == 0) {
+        continue;
+      }
+      stringCore::applyAppendersRecursive(out, [&](out_type<Varchar>& out) {
+        auto writeOffset = out.size();
+        out.resize(out.size() + value.size());
+        std::memcpy(out.data() + writeOffset, value.data(), value.size());
+      });
+    }
+
+    return true;
+  }
+};
+
+void registerVectorFunction() {
+  VELOX_REGISTER_VECTOR_FUNCTION(udf_concat, "concat_vector");
+}
+
+void registerSimpleFunctions() {
+  registerFunction<
+      StringConcatSimple,
+      Varchar,
+      Varchar,
+      Varchar,
+      Varchar,
+      Varchar,
+      Varchar>({"concat_simple"});
+
+  registerFunction<StringConcatVariadic, Varchar, Variadic<Varchar>>(
+      {"concat_simple_variadic"});
+}
+
+namespace {
+
+class VariadicBenchmark : public functions::test::FunctionBenchmarkBase {
+ public:
+  VariadicBenchmark() : FunctionBenchmarkBase() {
+    registerVectorFunction();
+    registerSimpleFunctions();
+  }
+
+  RowVectorPtr makeData(vector_size_t numArgs) {
+    const vector_size_t size = 1'000;
+
+    std::vector<VectorPtr> args;
+
+    for (vector_size_t i = 0; i < numArgs; ++i) {
+      args.emplace_back(
+          vectorMaker_.flatVector<StringView>(size, [](vector_size_t row) {
+            return StringView(fmt::format("{}", row));
+          }));
+    }
+
+    return vectorMaker_.rowVector(args);
+  }
+
+  size_t run(const std::string& functionName, vector_size_t numArgs = 5) {
+    folly::BenchmarkSuspender suspender;
+    auto rowVector = makeData(numArgs);
+
+    std::string args = "";
+    for (vector_size_t i = 0; i < numArgs; ++i) {
+      if (i > 0) {
+        args += ", ";
+      }
+
+      args += fmt::format("c{}", i);
+    }
+
+    auto exprSet = compileExpression(
+        fmt::format("{}({})", functionName, args), rowVector->type());
+    suspender.dismiss();
+
+    return doRun(exprSet, rowVector);
+  }
+
+  size_t doRun(exec::ExprSet& exprSet, const RowVectorPtr& rowVector) {
+    int cnt = 0;
+    for (auto i = 0; i < 100; i++) {
+      cnt += evaluate(exprSet, rowVector)->size();
+    }
+    return cnt;
+  }
+
+  bool hasSameResults(
```

Comment:
not related to this diff, 
we use this alot, might be nice to move it to FunctionBenchmarkBase if it can be generalized.